Update the char truncation section

aphillips opened the pull request:
Split from the work on #characters

~~NOTE WELL: this PR depends on edits done in #153, which have to be merged~~


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/aphillips/bp-i18n-specdev/pull/154.html" title="Last updated on Apr 3, 2025, 4:58 PM UTC (997a845)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/bp-i18n-specdev/154/0668d29...aphillips:997a845.html" title="Last updated on Apr 3, 2025, 4:58 PM UTC (997a845)">Diff</a>

aphillips committed:
Update the char truncation section

Split from the work on #characters

aphillips committed:
Address comments

aphillips committed:
Merge branch 'w3c:gh-pages' into char-truncation

aphillips committed:
Move family emoji example

aphillips committed:
Address comments from 2025-03-20 telecon

r12a commented on index.html:
```diff
@@ -3007,69 +3007,299 @@ <h3>Truncating or limiting the length of strings</h3>
 
-    <p>There are many reasons why a length limit might be needed in a specification or format. Generally length limits correspond to underlying limits in the implementation, such as the use of fixed-size fields in a database or data store, the desire to fit into practical boundaries such as packet size, or some other implementation detail related to storage allocation or efficiency.</p>
-
+    <p>There are many reasons why a length limit might be needed in a specification or format. The most common reason is because there is an underlying limit for storage or interchange, such as the use of a fixed-sized field in a database, the desire to fit into practical boundaries such as packet size, or some other implementation detail related to storage allocation or efficiency. The other com ...
```
```suggestion
    <p>There are many reasons why a length limit might be needed in a specification or format. The most common reasons are because there is an underlying limit for storage or interchange (such as the use of a fixed-sized field in a database), the desire to fit into practical boundaries such as packet size, or some other implementation detail related to storage allocation or efficiency. The other common reason is that there might be a limit on the size of the display or in the visible output.</p>
```

aphillips replied:
I made a slightly different change. I wanted to keep to "two reasons" and I think it is better to get away from the "there is a size limit" reason being an enumeration. See what you think of the revised version?

r12a commented on index.html:
```diff
@@ -3007,69 +3007,299 @@ <h3>Truncating or limiting the length of strings</h3>
 	<p class="advisement">Specifications SHOULD NOT limit the size of data fields unless there is a specific practical or technical limitation.</p>
```
The previous paragraph talks about "data structure or text field".

What's a 'data field'?  Is it a string in a database or a string presented to the reader? The text that follows appears to indicate the latter, since it refers to CSS text-overflow.  My first impression was the former.

aphillips committed:
Address @r12a's suggestion

r12a commented on index.html:
```diff
@@ -3007,69 +3007,299 @@ <h3>Truncating or limiting the length of strings</h3>
+        <p class="advisement">Avoid specifying data field length limits in terms of [=visual text units=] (such as [=grapheme clusters=]). Use mechanisms such as CSS <code>text-overflow</code> that depend on font metrics instead.</p>
+    </div>
+    
+    <p>If a text size limitation is related to the number of "display positions", it is tempting to want to use the number of [=visual text units=], since these generally correspond more closely to some arbitrary limit than the number of [=code points=] or [=code units=]. However, limits expressed in such units rarely accomplish their goal effectively because of the nature of text display. Propor ...
```
How about replacing 'display positions' (whatever those are) with:

> If a text-size limitation is related to the space available for displaying the text to the reader, it is tempting to want...

r12a commented on index.html:
```diff
@@ -3007,69 +3007,299 @@ <h3>Truncating or limiting the length of strings</h3>
     <p>When truncating strings, it's necessary to decide what units to use when counting the size of the string. In many cases this is beyond the control of the specification, since the truncation is occuring for some preordained reason. However, when the choice is available, some general guidelines can be applied.</p>
+    
+    <div class="req" id="char-trunc-avoid-visual-units">
+        <p class="advisement">Avoid specifying data field length limits in terms of [=visual text units=] (such as [=grapheme clusters=]). Use mechanisms such as CSS <code>text-overflow</code> that depend on font metrics instead.</p>
```
I don't think`text-overflow` has anything to do with font metrics.  It merely describes what happens to text that doesn't fit into the space available.

xfq replied:
https://drafts.csswg.org/css-overflow/#text-overflow

Yeah, I think font metrics is only one factor, and there are many other factors such as letter/word spacing etc.

r12a commented on index.html:
```diff
@@ -3007,69 +3007,299 @@ <h3>Truncating or limiting the length of strings</h3>
+        
+        <p>Tamil, like many South Asian scripts, is written in syllables. Each syllable is treated by Tamil speakers as a [=visual text unit=]. The word for "Unicode" in Tamil is a transliteration of the English word. If we divide the syllables up, as shown below, we can see that some are composed of multiple shapes and consume far more space than others:</p>
+        
+        <table>
```
I think the syllabification would normally be `juː.nik.koː.ʈu`.  Also, the pronunciation of the final syllable is probably 'ɖu'.  

And Google Translate translates it as யூனிகோட் — ie. `juː.ni.koːɖ`. Same for Bing.

Perhaps you meant 'orthographic syllables' ?

aphillips replied:
I got the translation from the old Unicode "what is unicode" pages. I can fix the syllables (I used a tool to generate the IPA... yours seems better)

r12a commented on index.html:
```diff
@@ -3007,69 +3007,299 @@ <h3>Truncating or limiting the length of strings</h3>
-    <p>If the limitation is related to the number of display positions, the grapheme count usually corresponds most closely to the expected limit. Note that proportional width fonts, combining marks, complex scripts, and many other factors complicate counting "screen positions". In Web pages, for example, the CSS <kbd>text-overflow</kbd> property provides visual truncation without disturbing the  ...
+    </aside>
+    
+    <p>Most limits are actually related to storage limitations, such as the size of a database field or a length limit in a protocol. Such limits are expressed in terms of [=code points=] in Unicode or [=code units=] (such as bytes) in a specific [=character encoding=]. Code points provides the best user experience, since all Unicode code points are treated identically: if text is truncated after ...
```
We seem to be missing some mustard here, along the lines of "When limits are related to storage limitations, express them using code points."

aphillips committed:
First bits from 2025-03-27 call (Tamil changes to follow)

aphillips committed:
Fixes to char_truncation per 2025-03-27 call.

r12a commented on index.html:
```diff
@@ -3004,109 +3004,347 @@ <h3>Truncating or limiting the length of strings</h3>
+    
+    <div class="req" id="char_trunc_units">
+	<div class="advisement"><p>Specifications that limit the length of a string MUST specify which type of unit the length limit uses. These might be:
+	   <ul>
```
As i suggested yesterday, i think we should drop this bulleted list because two of the bullets refer to things we don't recommend.  The mustard currently reads as if these are valid alternatives.  If you want to keep the text, i suggest adding it to the explanation below, introducing it using something like "Commonly used units for text segmentation include..."

r12a commented on index.html:
```diff
@@ -3004,109 +3004,347 @@ <h3>Truncating or limiting the length of strings</h3>
     <p>When truncating strings, it's necessary to decide what units to use when counting the size of the string. In many cases this is beyond the control of the specification, since the truncation is occuring for some preordained reason. However, when the choice is available, some general guidelines can be applied.</p>
+    
+    <div class="req" id="char-trunc-avoid-visual-units">
+        <p class="advisement">Avoid specifying string length limits in terms of [=visual text units=] (such as [=grapheme clusters=]). Use mechanisms such as CSS <code>text-overflow</code> instead.</p>
```
I think this needs to start out with "When the string length limitation is related to the size of the available display, avoid..."

aphillips replied:
I would have done that, but I thought this was a starker bit of advice. It isn't just when the length is related to display size.

r12a commented on index.html:
```diff
@@ -3004,109 +3004,347 @@ <h3>Truncating or limiting the length of strings</h3>
+            <tr>
+                <th>IPA</th>
+                <td style="width:6rem;text-align:center">juː</td>
+                <td style="width:6rem;text-align:center">.ni</td>
```
The dot (syllable separator in IPA) belongs after a syllable, rather than before. So juː. ni. etc.

aphillips replied:
🤦‍♂️

r12a commented on index.html:
```diff
@@ -3004,109 +3004,347 @@ <h3>Truncating or limiting the length of strings</h3>
 	<p class="advisement">Specifications that limit the length of a string SHOULD specify the length in terms of Unicode code points.</p>
 	</div>
+    
+    <p>Most limits are actually related to storage limitations, such as the size of a database field or a length limit in a protocol. Such limits are expressed in terms of [=code points=] in Unicode or [=code units=] (such as bytes) in a specific [=character encoding=]. Code points provides the best user experience, since all Unicode code points are treated identically: if text is truncated after ...
```
I suspect that the danger we're trying to head off here is that implementers tend to, by default, suppose that they can set storage limitations that are based on a number of bytes, regardless of encodings.  So what we're trying to convince them of is that simple byte counts break things and so the encoding units need to be taken into account — even though this means that the actual storage space taken up (in bits & bytes) is going to be rather fuzzy.

I suspect that it would be good to say that in the explanation immediately after the first sentence, and then to carry on with "Such limits are better expressed in terms of ..."

Note that we may need, however, to distinguish between counting plain bytes vs counting code units (which may equate to bytes in some cases).  I'm not sure that that's clear here.

aphillips replied:
I agree. We also don't talk anywhere about "inflation factors". That is, if you want to guarantee that a byte-length measured field will store 50 code points in UTF-8, you need to make it 50 * (max_size_of_encoding) in length, or 200 bytes long.

r12a commented on index.html:
```diff
@@ -3004,109 +3004,347 @@ <h3>Truncating or limiting the length of strings</h3>
+	</div>
+    
+    <div class="req" id="char_trunc_indicator">
+	   <p class="advisement">If a specification specifies a length limit, it SHOULD specify that any string that has been truncated include an indicator, such as ellipses, that the string has been altered.</p>
```
```suggestion
	   <p class="advisement">If a specification specifies a length limit, it SHOULD specify that any string that has been truncated include an indicator that the string has been altered, such as an ellipsis.</p>
```

aphillips replied:
```suggestion
<p class="advisement">If a specification specifies a length limit, it SHOULD specify that any string that has been truncated includes an indicator that the string has been altered, such as an ellipsis.</p>
```

r12a commented on index.html:
```diff
@@ -3004,109 +3004,347 @@ <h3>Truncating or limiting the length of strings</h3>
+ 
+    
 	<div class="req" id="char_trunc_unit_rec">
 	<p class="advisement">Specifications that limit the length of a string SHOULD specify the length in terms of Unicode code points.</p>
```
```suggestion
	<p class="advisement">Specifications that limit the maximum allowable length of a string SHOULD specify the length in terms of Unicode code points.</p>
```

This may help to make it clear that we're talking about container lengths here, rather than truncation points.

r12a commented on index.html:
```diff
@@ -3004,109 +3004,347 @@ <h3>Truncating or limiting the length of strings</h3>
+	</div>
+    
+    <aside style="border:1px solid green; padding: 5px" title="Important">
+        <p><q><em>Truncation on [=visual text unit=] boundaries</em></q> is not the same thing as using [=visual text units=] as the length limit.</p>
```
```suggestion
        <p><q><em>Truncation on [=visual text unit=] boundaries</em></q> is not the same thing as using [=visual text units=] as the maximum length limit for the container.</p>
```

r12a commented on index.html:
```diff
@@ -3004,109 +3004,347 @@ <h3>Truncating or limiting the length of strings</h3>
+            </tr>
+            <tr>
+                <th>IPA</th>
+                <td style="width:6rem;text-align:center">juː</td>
```
Move the dots again.

r12a replied:
If you want a word that contains both an inherent vowel and 'ko' in Tamil you could use this:

நன்கொடை
n̪an.ɡo.ɖai
donation

I found this by searching at https://r12a.github.io/scripts/taml/ta_vocab.html?q=%E2%92%B8%E2%92%B8.*%E0%AE%95%E0%AF%8A

See its composition at https://r12a.github.io/pickers/taml/?text=%E0%AE%A8%E0%AE%A9%E0%AF%8D%E0%AE%95%E0%AF%8A%E0%AE%9F%E0%AF%88 (click on List Characters)

Wiktionary entry is https://en.wiktionary.org/wiki/%E0%AE%A8%E0%AE%A9%E0%AF%8D%E0%AE%95%E0%AF%8A%E0%AE%9F%E0%AF%88#Tamil

r12a replied:
Aonther would be the following. This has no inherent vowel, but has a one code point initial vowel letter.

ஆலக்கொடி
aː.lak.ko.ɖi
yellow orpiment

https://en.wiktionary.org/wiki/%E0%AE%86%E0%AE%B2%E0%AE%95%E0%AF%8D%E0%AE%95%E0%AF%8A%E0%AE%9F%E0%AE%BF#Tamil

aphillips replied:
TIL what 'yellow orpiment' is. 😸 

I don't really want a random word, if I can help it. I like using the word Unicode, in part because it is a transliterated word that should be well-known to spec writers (and fits in our world). It doesn't introduce every facet of Unicode. The only downside is that it introduces an urge to explain that Tamil doesn't always use pairs of code points. I've thought of a better way to explain this, which I'll add.

aphillips replied:
This is the one change I didn't make, but I wrote a ton of new BPs and did more writing/reorg. Time for another look?

r12a commented on index.html:
```diff
@@ -3004,109 +3004,347 @@ <h3>Truncating or limiting the length of strings</h3>
+      
+        <p>If we arbitrarily truncate this text after the 16th byte (where the yellow background begins above), we not only divide a [=grapheme cluster=], but also one of the [=code points=]. The resulting damaged character is replaced in a UTF-8 decoder with the <span class="codepoint"><bdi lang="en">&#xfffd;</bdi><code class="uname">U+FFFD REPLACEMENT CHARACTER</code></span>. In [=legacy charac ...
 
+        <p class="bigtext" lang="ta">யூனிக�...</p>
```
I'm not seeing big text — i assume that's the fault of the preview application.

aphillips replied:
No... might be your browser?? LGTM in FF and Chrome on Windoze and FF on Linux. If it doesn't work on Safari, might need to look at the styles or images.

![image](https://github.com/user-attachments/assets/091bfdc0-000d-441c-9250-d26e231e700e)

r12a commented on index.html:
```diff
@@ -3004,109 +3004,347 @@ <h3>Truncating or limiting the length of strings</h3>
+	<p>When choosing a length limit, bear in mind the needs of different scripts and languages in Unicode. If a text size limit is set in terms of [=code units=] (such as a byte-length limit) in a given [=character encoding=], the limit needs to consider the relative efficiency of the character encoding for different scripts. In particular, UTF-8 is the multibyte character encoding of [[Unicode]]. U ...
+    
+    		
+	<p>When the limit placed on a text field is counted in bytes, the number of characters that can be stored in that field depends on which characters are being stored. To ensure that users who speak various languages are not at a disadvantage, the limit needs to allow for a reasonable number of charactes as encoded in that language.</p>
```
```suggestion
	<p>When the limit placed on a text field is counted in bytes, the number of characters that can be stored in that field depends on which characters are being stored. To ensure that users who speak various languages are not at a disadvantage, the limit needs to allow for a reasonable number of characters as encoded in that language.</p>
```

r12a reviewed (commented):
I don't remember seeing any mention of the dangers of splitting surrogate pairs when working with code units, which surprises me.  I'd have thought that that is one of the key issues to avoid?

aphillips committed:
Update index.html

aphillips committed:
Update index.html

Co-authored-by:   r12a <ishida@w3.org>

aphillips committed:
Update index.html

Co-authored-by:   r12a <ishida@w3.org>

aphillips committed:
Update index.html

Co-authored-by:   r12a <ishida@w3.org>

aphillips committed:
Address comments

r12a commented on index.html:
```diff
@@ -3001,112 +3001,359 @@ <h3>Truncating or limiting the length of strings</h3>
 
 
-<p>Some specifications, formats, or protocols or their implementations need to specify limits for the size of a given data structure or text field. This could be due to many reasons, such as limits on processing, memory, data structure size, and so forth. When selecting or specifying limits on the length of a given string, specifications or implementations need to ensure that they do not cause co ...
+<p>Some specifications, formats, or protocols or their implementations need to specify limits for the size of a given string. This could be due to many reasons, such as limits on processing, memory, data structure size, and so forth. Because length limits often involve truncating or limiting the size of text, specification and their implementations need to apply extra care to ensure that the text ...
```
```suggestion
<p>Some specifications, formats, or protocols or their implementations need to specify limits for the size of a given string. This could be due to many reasons, such as limits on processing, memory, data structure size, and so forth. Because length limits often involve truncating or limiting the size of text, specifications and their implementations need to apply extra care to ensure that the text is not corrupted and that the limit chosen does not make the feature unusable for certain audiences.</p>
```

r12a commented on index.html:
```diff
@@ -3001,112 +3001,359 @@ <h3>Truncating or limiting the length of strings</h3>
+    <p>There are many reasons why a length limit might be needed in a specification or format. The most common reason is because there is an underlying size limit on the data. For example, there might be a fixed-sized field in a database, a practical boundary such as packet size, or some other implementation detail related to storage allocation or efficiency. The other common reason is that there ...
+    
+    <div class="req" id="char_trunc_units">
+	   <p class="advisement">Specifications that limit the length of a string MUST specify which whether the limit is counted in Unicode [=code points=] or is counted in [=code units=] (such as bytes) of a given [=character encoding=].</p>
```
```suggestion
	   <p class="advisement">Specifications that limit the length of a string MUST specify whether the limit is counted in Unicode [=code points=], or is counted in [=code units=] (such as bytes) of a given [=character encoding=].</p>
```

r12a commented on index.html:
```diff
@@ -3001,112 +3001,359 @@ <h3>Truncating or limiting the length of strings</h3>
+	   <p class="advisement">Specifications that limit the length of a string MUST specify which whether the limit is counted in Unicode [=code points=] or is counted in [=code units=] (such as bytes) of a given [=character encoding=].</p>
+	</div>
+    
     <p>When truncating strings, it's necessary to decide what units to use when counting the size of the string. In many cases this is beyond the control of the specification, since the truncation is occuring for some preordained reason. However, when the choice is available, some general guidelines can be applied.</p>
```
```suggestion
    <p>When truncating strings, it's necessary to decide what units to use when counting the size of the string. In some cases this is beyond the control of the specification, since the truncation is occuring for some preordained reason. However, when the choice is available, some general guidelines can be applied.</p>
```

r12a commented on index.html:
```diff
@@ -3001,112 +3001,359 @@ <h3>Truncating or limiting the length of strings</h3>
     <p>When truncating strings, it's necessary to decide what units to use when counting the size of the string. In many cases this is beyond the control of the specification, since the truncation is occuring for some preordained reason. However, when the choice is available, some general guidelines can be applied.</p>
+    
+    <div class="req" id="char-trunc-avoid-visual-units">
+        <p class="advisement">Avoid specifying string length limits in terms of [=visual text units=] (such as [=grapheme clusters=]). Use mechanisms such as CSS <code>text-overflow</code> [[css-overflow-4]] instead.</p>
```
```suggestion
        <p class="advisement">Avoid specifying string length limits in terms of [=visual text units=] (such as [=grapheme clusters=]). If a string length limitation is related to the size of the available display, use mechanisms such as CSS <code>text-overflow</code> [[css-overflow-4]] instead.</p>
```

r12a replied:
Actually, given the explanation in the following paragraph, i still think that this mustard really should relate to visual truncation only — in which case, move the suggested addition to the beginning of the mustard.

aphillips replied:
A good bit of work went into the next iteration. See what you think.

r12a commented on index.html:
```diff
@@ -3001,112 +3001,359 @@ <h3>Truncating or limiting the length of strings</h3>
-			<td><code>F0 9F 99 8A F0 9F 99 81 F0 9F 98 A2 F0 9F 98 A0 F0 9F 98 A7 F0 9F 98 8E F0 9F 98 BD F0 9F 98 89 F0 9F 98 84 F0 9F 98 AE</code></td>
-		</tr>
-		</table>
+        <p>Finally, notice that the Tamil word takes more room with four [=visual text units=] than an English word with nine:</p>
```
It doesn't take up more space in the page i'm looking at.  I think you need an image here to make your point, and you need to more carefully manage the relative font sizes.

<img width="300" alt="Screenshot 2025-04-03 at 11 39 40" src="https://github.com/user-attachments/assets/f28fd71a-6ff6-4c57-8d50-f938ee58e16c" />

aphillips replied:
done

r12a commented on index.html:
```diff
@@ -3001,112 +3001,359 @@ <h3>Truncating or limiting the length of strings</h3>
+    
 	<div class="req" id="char_trunc_unit_rec">
-	<p class="advisement">Specifications that limit the length of a string SHOULD specify the length in terms of Unicode code points.</p>
+	<p class="advisement">Specifications that limit the maximum allowable length of a string SHOULD specify the length in terms of [=Unicode code points=].</p>
```
Unless you're using CSS `text-overflow`... Again, i think we need to more clearly separate and signpost whether we're talking about storage vs presentational truncation, because they require different approaches and mustard.

aphillips replied:
Here I don't quite agree. The requirement is about the "maximum allowable length of a string", i.e. the stored string. To make it clear, I changed it to "maximum allowable storage length"

r12a commented on index.html:
```diff
@@ -3001,112 +3001,359 @@ <h3>Truncating or limiting the length of strings</h3>
-	<div class="req" id="char_trunc_byte_boundary">
-	<p class="advisement">If a specification sets a length limit in code units (such as bytes), it MUST specify that truncation can only occur on code point boundaries.</p>
+    
+    <p>Most limits are actually related to storage limitations, such as the size of a database field or a length limit in a protocol. Such limits are expressed either in terms of [=code points=] in Unicode or in terms of [=code units=] (such as bytes) in a specific [=character encoding=]. [=Code points=] provides the best user experience, since all Unicode code points are treated identically: if  ...
```
```suggestion
    <p>Most limits are actually related to storage limitations, such as the size of a database field or a length limit in a protocol. Such limits are expressed either in terms of [=code points=] in Unicode, or they are expressed in terms of [=code units=] (such as bytes) in a specific [=character encoding=]. [=Code points=] provide the best user experience, since all Unicode code points are treated identically: if text is truncated after 40 code points, all languages and scripts get the same number of code points to work with. By contrast, when the size limit is expressed in [=code units=], such as bytes in UTF-8, users who write in a language that mostly uses ASCII letters get many more characters (code points) for a given size limit than user's whose language is mostly made up of characters that take 2-, 3-, or 4-bytes per code point.</p>
```

r12a replied:
The first proposed additions are to clarify that the specific encoding is only relevant to the code point approach.

r12a commented on index.html:
```diff
@@ -3001,112 +3001,359 @@ <h3>Truncating or limiting the length of strings</h3>
+    <p>Most limits are actually related to storage limitations, such as the size of a database field or a length limit in a protocol. Such limits are expressed either in terms of [=code points=] in Unicode or in terms of [=code units=] (such as bytes) in a specific [=character encoding=]. [=Code points=] provides the best user experience, since all Unicode code points are treated identically: if  ...
+    
+    <div class="req" id="char_trunc_grapheme_boundary">
+	   <p class="advisement">Specifications that limit the length of a string SHOULD require truncation on [=visual text unit=] boundaries (usually as approximated by [=grapheme cluster=] boundaries). Truncation in the middle of a visual text unit such as a <a>combining character sequence</a> can alter the meaning of the string.</p>
```
```suggestion
	   <p class="advisement">Specifications that require truncation of strings SHOULD truncate the text at [=visual text unit=] boundaries (usually as approximated by [=grapheme cluster=] boundaries). Truncation in the middle of a visual text unit such as a <a>combining character sequence</a> can alter the meaning of the string.</p>
```

r12a replied:
The second sentence "Truncation in the middle..." is actually an explanation, so should be dropped into the start of the explanatory text that follows, rather than included in the mustard.

r12a replied:
This was as far as i could get with my review for this week.

aphillips replied:
I did a deeper revision, in part to address the second comment.

aphillips committed:
Apply suggestions from code review

Co-authored-by:   r12a <ishida@w3.org>

aphillips committed:
Address comments
- Add image for Tamil 'unicode' vs. 'memorable'
- Redo the mustard for visual truncation and associated text
